review/jm-initial

fivetran-joemarkiewicz opened the pull request:
This is a PR that is meant to serve as the initial review of the Qualtrics package. The notes included in this review are intended to be addressed by being made directly to the base branch. This PR is not intended to be merged.

fivetran-joemarkiewicz commented on dbt_project.yml:
```diff
@@ -11,7 +11,7 @@ models:
     +materialized: table
     intermediate:
-      +materialized: view
+      +materialized: view 
```
What was the reasoning to materialize these as views opposed to ephemeral? I am open to views but more generally curious.

fivetran-jamie replied:
ah just for testing -- switched to ephemeral!

fivetran-joemarkiewicz commented on models/intermediate/int_qualtrics__responses.sql:
```diff
@@ -44,7 +44,7 @@ response_join as (
     
     from question_response 
-    join survey_response 
+    join survey_response
```
Can we explicitly call out the join we are intending to do here? If this is meant to be an inner join can we explicitly add that.

fivetran-jamie replied:
updated to inner

fivetran-joemarkiewicz commented on models/intermediate/int_qualtrics__survey.sql:
```diff
@@ -62,7 +62,7 @@ survey_join as (
         on survey.creator_user_id = creator.user_id
         and survey.source_relation = creator.source_relation
-    join agg_versions 
+    join agg_versions
```
Same comment about explicitly defining the type of join.

fivetran-jamie replied:
updated to left

fivetran-joemarkiewicz commented on models/qualtrics__contact.sql:
```diff
@@ -40,7 +40,7 @@ distribution_response as (
 
     from distribution_contact 
-    join distribution 
+    join distribution
```
Same comment about explicitly defining the type of join.

fivetran-jamie replied:
updated to left

fivetran-joemarkiewicz commented on models/qualtrics__contact.sql:
```diff
@@ -54,8 +54,8 @@ agg_distribution_responses as (
         source_relation,
-        -- should i include other distribution channels? qr code, social media, personal links, etc?
-        -- should i add avg time-to-open?
+        -- should i include other distribution channels? qr code, social media, personal links, etc? 
```
I believe we landed on not needing to include these. If that is the case, feel free to remove the comment.

fivetran-jamie replied:
removed!

fivetran-joemarkiewicz commented on models/qualtrics__contact.sql:
```diff
@@ -54,8 +54,8 @@ agg_distribution_responses as (
-        -- should i include other distribution channels? qr code, social media, personal links, etc?
-        -- should i add avg time-to-open?
+        -- should i include other distribution channels? qr code, social media, personal links, etc? 
+        -- should i add avg time-to-open? 
```
I think that would be an interesting metric to gauge how active a contact is. What is the LOE to add that in? If it is not too much I say include it, if it is a large lift then let's exclude for now as it is not crucial.

fivetran-jamie replied:
ah yeah i'll just go ahead and remove

fivetran-joemarkiewicz commented on models/qualtrics__contact.sql:
```diff
@@ -68,7 +68,7 @@ agg_distribution_responses as (
 ),
 
--- todo: add medians
+-- todo: add medians 
```
Were you still planning to include medians in this model?

fivetran-jamie replied:
added!

fivetran-joemarkiewicz commented on models/qualtrics__directory.sql:
```diff
@@ -68,10 +68,10 @@ agg_distributions as (
         
     from distribution_contact
-    join directory_contact 
+    join directory_contact
```
Same comment about explicitly defining the type of join.

fivetran-jamie replied:
updated to inner

fivetran-joemarkiewicz commented on models/qualtrics__directory.sql:
```diff
@@ -68,10 +68,10 @@ agg_distributions as (
         on distribution_contact.contact_id = directory_contact.contact_id
         and distribution_contact.source_relation = directory_contact.source_relation
-    join distribution 
+    join distribution
```
Same comment about explicitly defining the type of join.

fivetran-jamie replied:
updated to left

fivetran-joemarkiewicz commented on models/qualtrics__directory.sql:
```diff
@@ -88,13 +88,13 @@ final as (
-        agg_distributions.count_contacts_started_survey_30d,
-        agg_distributions.count_contacts_completed_survey_30d,
-        agg_distributions.count_surveys_sent_30d,
+        agg_contacts.count_contacts_created_30d, 
+        agg_contacts.count_contacts_unsubscribed_30d, 
+        agg_distributions.count_contacts_sent_survey_30d, 
+        agg_distributions.count_contacts_opened_survey_30d, 
+        agg_distributions.count_contacts_started_survey_30d, 
+        agg_distributions.count_contacts_completed_survey_30d, 
+        agg_distributions.count_surveys_sent_30d, 
```
A number of these are showing as null in our data. Can we coalesce to 0. Or do you feel it could cause confusion if we are saying `0` when there has actually been no activity 🤔. What are your thoughts?

fivetran-jamie replied:
ah yeah meant to coalesce everything! except for avg's and medians

fivetran-joemarkiewicz commented on models/qualtrics__distribution.sql:
```diff
@@ -57,20 +57,20 @@ pivoted_metrics as (
-        {% if target.type == 'postgres' %} group by 1,2 {% endif %}
-    )
-    {% if target.type != 'postgres' %} group by 1,2,3,4,5 {% endif %}
+calc_medians as ( 
+     
+    select * from ( 
+        select  
+            distribution_id,  
+            source_relation, 
+            {{ fivetran_utils.percentile(percentile_field='time_to_open_in_seconds', partition_field='distribution_id,source_relation', percent='0.5') }} as median_time_to_open_in_seconds, 
+            {{ fivetran_utils.percentile(percentile_field='time_to_start_in_seconds', partition_field='distribution_id,source_relation', percent='0.5') }} as median_time_to_start_in_seconds, 
+            {{ fivetran_utils.percentile(percentile_field='time_to_complete_in_seconds', partition_field='distribution_id,source_relation', percent='0.5') }} as median_time_to_complete_in_seconds 
+ 
+        from distribution_contact 
+        {% if target.type == 'postgres' %} group by 1,2 {% endif %} 
+    ) 
+    {% if target.type != 'postgres' %} group by 1,2,3,4,5 {% endif %} 
```
I am not fully following this cte. Would you be able to explain the need for the subquery and why postgres needs to be treated differently?

fivetran-jamie replied:
yeah so for postgres, this [macro](https://github.com/fivetran/dbt_fivetran_utils/tree/v0.4.7#percentile-source) uses an aggregate function, whereas it is a window function on other databases.  i have the second `group by` in order to roll up everything to the same grain (essentially mimicking an aggregate function)

fivetran-jamie replied:
added some in-line comments

fivetran-joemarkiewicz replied:
Ahhh okay thanks for the explanation and also the inline comments. Thanks!

fivetran-joemarkiewicz commented on models/qualtrics__response_breakdown.sql:
```diff
@@ -1,5 +1,5 @@
 with response as (
-    -- will probably move the logic from this int model to the end model directly 
+    -- will probably move the logic from this int model to the end model directly  
```
I think it is fine to keep as intermediate model

fivetran-jamie replied:
cool yah if int models are ephmeral this doesn't matter too much

fivetran-joemarkiewicz commented on models/qualtrics__survey.sql:
```diff
@@ -16,7 +16,7 @@ question as (
 ),
 
--- should we bring in distribution_contact for like the number of currently pending or opened invites?
+-- should we bring in distribution_contact for like the number of currently pending or opened invites? 
```
Similar comment around this being a good metric that I feel could add a lot of value, but not necessarily crucial in my opinion. If the LOE is relatively small then I say yes. If not, then I say we hold off for now.

fivetran-jamie replied:
removing

fivetran-joemarkiewicz commented on models/qualtrics__survey.sql:
```diff
@@ -73,21 +73,21 @@ agg_responses as (
-    )
-    {% if target.type != 'postgres' %} group by 1,2,3,4 {% endif %}
-),
+calc_medians as ( 
+ 
+    select * from ( 
+ 
+        select  
+            survey_id,  
+            source_relation, 
+            {{ fivetran_utils.percentile(percentile_field='duration_in_seconds', partition_field='survey_id,source_relation', percent='0.5') }} as median_response_duration_in_seconds, 
+            {{ fivetran_utils.percentile(percentile_field='progress', partition_field='survey_id,source_relation', percent='0.5') }} as median_survey_progress_pct 
+ 
+        from responses 
+        {% if target.type == 'postgres' %} group by 1,2 {% endif %} 
+    ) 
+    {% if target.type != 'postgres' %} group by 1,2,3,4 {% endif %} 
+), 
```
Same comment about explaining in a bit more detail.

fivetran-joemarkiewicz reviewed (commented):
@fivetran-jamie fantastic work on this PR!! I see this providing A LOT of value to Qualtrics users. It solves a lot of my main pain points with trying to analyze survey data. Also, great work on developing the models. I can tell this was quite complex, but I found very little questions or issue with your logic and am very impressed with the level of detail that went into this package build. Great job!

I do have a few questions below that I would like your eyes on or to make updates. Additionally I had a minor question not included below:
- Should we include `_overview` and/or `_summary` to the non breakdown models or do you feel that is duplicative?

fivetran-joemarkiewicz commented:
Thanks for addressing the comments @fivetran-jamie! This looks great to me. I will open a final PR to discuss any notes around the docs and daily model.
